### PR TITLE
Alish Term proposal

### PR DESCRIPTION
**Pull request by alishsuper**

### CODE REVIEW

File: term_proposal/Term_proposal.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+## Why do I want to do this project?
+In my last Kaggle competition related to an image classification problem, the victory was won by a girl from Japan who used XGBoost algorithm. Even this algorithm is very popular, the most people just try to tune hyperparameters for training the model. But very few people implement XGBoost by their own. The reason is obvious, if a powerful enough tool already exists, then why come up with something new.
+However, I think that for education process is really helpful when you don’t use libraries, but try to implement it by yourself. It can give more insight into how XGBoost works.
```
Comment (yungyuc) [inline term_proposal/Term_proposal.md:10]: I agree implementing a successful tool from ground up will help you a lot for developing skills in computational science.  It's also something you don't always have the luxury to do in a research project.  And it's certainly the purpose of the term project of this course.  I don't know about XGBoost, but if my understanding (from your description) is correct, it's something related to neural network and/or linear algebra with the application of machine learning, and it fits my expectation of a numerical software project.

I'd hope you can be more specific about what you'd like to practice (the educational purpose) in the proceeding sections.  I see you made some good starting points and will leave concrete suggestions accordingly.  Here I'd like you to think about in a global sense.  You should state in the proposal what do you expect to learn from this experience.

File: term_proposal/Term_proposal.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+I propose a software project that implements small part of XGBoost algorithm related to regression problem.
+
+## What am I going to do in this project?
+Learn and understand the math background of XGBoost algorithm.
```
Comment (yungyuc) [inline term_proposal/Term_proposal.md]: If you have not a sufficient understanding to the algorithm, it's risky.  I suggest you to outline what you will expect from the algorithm and add estimation of time and stage the learning.  As a rule of thumb, to learn what you want to code in the project shouldn't take more than 3 weeks.

File: term_proposal/Term_proposal.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+Implement XGBoost algorithm using linear booster (or if there is enough time, then also gradient booster).
+Plot all results to get better understanding of the XGBoost algorithm.
```
Comment (yungyuc) [inline term_proposal/Term_proposal.md]: Specify what types of plotting you'd like to do.  If there is a priority, list them.

File: term_proposal/Term_proposal.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+Write unit tests.
```
Comment (yungyuc) [inline term_proposal/Term_proposal.md]: Specify how you'd plan to test.

File: term_proposal/Term_proposal.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+Compare to the original XGBoost.
```
Comment (yungyuc) [inline term_proposal/Term_proposal.md]: You can group this to unit tests, or use the unit tests framework for this (functional or acceptance) test.  Essentially it's validating your implementation.

File: term_proposal/Term_proposal.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+Prepare final report and presentation.
+
+## References.
+
+- https://www.kaggle.com/
+- https://github.com/dmlc/xgboost
+- https://xgboost.readthedocs.io/en/latest/
```
Comment (yungyuc) [inline term_proposal/Term_proposal.md:33]: You can analyze the architecture of the existing tool, and state how you'd like to learn and improve it.

**Review by yungyuc — CHANGES_REQUESTED** [review]

File: term_proposal/Term_proposal.md (hunk 1 above)
Comment (alishsuper) [reply term_proposal/Term_proposal.md:10]: Dear Yung-Yu. Your understanding is completely correct. The purpose of this project is better understanding of this method. Because now I just know how to apply it, but I don't understand why XGBoost works so well. So I hope that implementing small part of this powerful tool gives me good understanding of this.

File: term_proposal/Term_proposal.md (hunk 2 above)
Comment (alishsuper) [reply term_proposal/Term_proposal.md]: Sure Sir. I will add this information. Actually XGBoost is huge project so it is impossible to implement everything. I think that I will just consider the small part for particular task (linear regression).

File: term_proposal/Term_proposal.md (hunk 3 above)
Comment (alishsuper) [reply term_proposal/Term_proposal.md]: 1. Approximation by XGBoost on test data
2. Loss function on train and test data

File: term_proposal/Term_proposal.md (hunk 4 above)
Comment (alishsuper) [reply term_proposal/Term_proposal.md]: I thought that I can use unit tests to test modules of my program, for example, to avoid null or infinity results.
Building XGBoost model includes these steps:
1. Prepossessing.
2. Training.
3. Prediction.
4. Visualization. 
I am going to write unit tests at each of these steps for reliable program operation.

File: term_proposal/Term_proposal.md (hunk 5 above)
Comment (alishsuper) [reply term_proposal/Term_proposal.md]: Can I use original XGBoost just to compare final results? I also want to visualize the results of original XGBoost.

File: term_proposal/Term_proposal.md (hunk 6 above)
Comment (alishsuper) [reply term_proposal/Term_proposal.md:33]: I am not sure that I can improve the architecture, otherwise it could be one very good paper. I think the main goal of my project is to study and implement this tool based on math background. I am not going to look at the original code, so I will not know how they implemented original XGBoost. I just have the architecture and I will try to use it.

File: term_proposal/Term_proposal.md (hunk 2 above)
Comment (yungyuc) [reply term_proposal/Term_proposal.md]: Your estimate of the size of XGBoost is a good piece of information, pal.  Being specific certainly reduces the risk in the project.

File: term_proposal/Term_proposal.md (hunk 4 above)
Comment (yungyuc) [reply term_proposal/Term_proposal.md]: Visualization is especially difficult to test.  If you don't insist, I'd suggest you to not put too much efforts in it.  Some testing is good, still.

File: term_proposal/Term_proposal.md (hunk 5 above)
Comment (yungyuc) [reply term_proposal/Term_proposal.md]: That sounds like a plan.

File: term_proposal/Term_proposal.md (hunk 6 above)
Comment (yungyuc) [reply term_proposal/Term_proposal.md:33]: Improving the algorithm can probably be a good research paper.  Improving the implementation, on the other hand, often costs one's tenure.  What I meant was the latter.

I did suggest you to take a look at the original code, but it is not a strong opinion.  If this is your life-long work, I'd say you must read as much code as you can.  But in the setting of a term project, use your own judgement.

**Comment by yungyuc** [comment]
I am taking this proposal and merging the PR, although my requests weren't all changed in the document.  You may track them yourself, @alishsuper, or file another PR to do the update.  This proposal can be considered as a design guideline for your project, and having it comprehensive helps you.
